Source/ProgressionSystemRuntime/Private/Widgets/PSMenuWidget.cpp

```diff
@@ -0,0 +1,83 @@
+﻿// Copyright (c) Valerii Rotermel & Yevhenii Selivanov
+
+#include "Widgets/PSMenuWidget.h"
+//---
+
+//---
+#include "Data/PSDataAsset.h"
+#include "Blueprint/WidgetTree.h"
+#include "Components/HorizontalBox.h"
+#include "Components/HorizontalBoxSlot.h"
+#include "Components/Image.h"
+#include "Data/PSWorldSubsystem.h"
+
+#include UE_INLINE_GENERATED_CPP_BY_NAME(PSMenuWidget)
+
+void UPSMenuWidget::AddImagesToHorizontalBox(int32 AmountOfUnlockedPoints, int32 AmountOfLockedPoints)
+{
+	checkf(HorizontalBox, TEXT("ERROR: 'HorizontalBox' is null"));
+	if (HorizontalBox)
+	{
+		// Loop to create and add images to the Horizontal Box for unlocked stars
+		for (int32 i = 0; i < AmountOfUnlockedPoints; i++)
+		{
+			UImage* ImageWidget = WidgetTree->ConstructWidget<UImage>(UImage::StaticClass());
+			ImageWidget->SetBrushFromTexture(UPSDataAsset::Get().GetUnlockedProgressionIcon());
+
+			// Load and set the image texture here using ImagePath or other methods
+			HorizontalBox->AddChildToHorizontalBox(ImageWidget);
+
+			// Create the UHorizontalBoxSlot and assign it to the ImageWidget
+			UHorizontalBoxSlot* HorizontalBoxSlot = Cast<UHorizontalBoxSlot>(ImageWidget->Slot);
+			if (HorizontalBoxSlot)
+			{
+				HorizontalBoxSlot->SetPadding(FMargin(20.0f, 10.0f, 20.0f, 10.0f));
+				HorizontalBoxSlot->SetSize(FSlateChildSize(ESlateSizeRule::Fill));
+				HorizontalBoxSlot->SetHorizontalAlignment(HAlign_Fill);
+				HorizontalBoxSlot->SetVerticalAlignment(VAlign_Fill);
+			}
+		}
+
+		// Loop to create and add images to the Horizontal Box for unlocked stars
+		for (int32 i = 0; i < AmountOfLockedPoints; i++)
+		{
+			UImage* ImageWidget = WidgetTree->ConstructWidget<UImage>(UImage::StaticClass());
```

Comment:
The construction of widgets inside loops could be a significant performance hit, especially if created and destroyed more than once. I see you do it in `UpdateProgressionWidgetForPlayer` that happens everytime player enters to Menu or switches the player.

**It is very expensive** to call such often `ConstructWidget<UImage>()` following with `ClearChildren()`.

Please optimize it with A or B:

- A) proper way.
Use 'Pool Manager` plugin that is already enabled in Bomber for pooling widgets instead of constructing\destroying new ones every time. Widget pooling will drastically reduce the overhead of dynamic allocation and destruction, especially in a loops. You can find its example in `AGeneratedMap::SpawnActorsByTypes`: `UPoolManagerSubsystem::Get().TakeFromPool`.

Pros of solution A is the pooling logic is already implemented there, you would not need to manage any logic on Bomber side at all, just call `TakeFromPool` and `ReturnToPool`. That's all

Cons of solution A is that I did not implement widgets support for Pool Manger. You would need to extend Pool Manager plugin to support `UUserWidget` class. To do so, new `PoolFactory_Widget` factory is needed to be created on Pool Manager side with own pull request. But it shouldn't complicated at all. See [doc how to create new factory](https://docs.google.com/document/d/1YxbIdc9lZRl5ozI7_1LTBfdzJWTwhxwz2RKA-r0Q4po/edit#heading=h.5r9wwe41vypo).

- B) easiest way
Or alternative solution is precreate in your PSMenuWidget all those subwidgets in `NativeOnInitialized` function (begin play of widgets that is called only once) based.

Cons of solution B is that you will need to proper manage caching of such subwidgets. It could be difficult since number of stars could be different.

Reply:
Pool manager added: https://github.com/h4rdmol/ProgressionSystem/pull/1/commits/13240533c4f5e93d8deb442087af6787233556ce

Reply:
fixed issues with the pool manager https://github.com/h4rdmol/ProgressionSystem/pull/1/commits/de098d441b250c377a9e1198c0eab50446afbfdf

Reply:
resolved open todo's as part of the pool manager fixes: https://github.com/h4rdmol/ProgressionSystem/pull/1/commits/d1a2f130d516629c9dcf7de6326828492d107e01